pyacq/core/sharedarray.py

```diff
@@ -55,13 +43,62 @@ def __init__(self, shape=(1,), dtype = 'float64', shm_id = None):
                 self.mmap = mmap.mmap(self._tmpFile.fileno(), self.nbytes, mmap.MAP_SHARED, mmap.PROT_READ)
                 
     def close(self):
+        """Close this buffer.
+        """
         self.mmap.close()
         if not sys.platform.startswith('win') and hasattr(self, '_tmpFile'):
             self._tmpFile.close()
     
     def to_dict(self):
-        return {'shape': self.shape, 'dtype': self.dtype, 'shm_id': self.shm_id}
+        """Return a dict that can be serialized and sent to other processes to
+        access this buffer.
+        """
+        return {'nbytes': self.nbytes, 'shm_id': self.shm_id}
+    
+    def to_numpy(self, offset, dtype, shape, strides):
+        """Return a numpy array pointing to part (or all) of this buffer.
+        """
+        return np.ndarray(buffer=self.mmap, shape=shape,
+                          strides=strides, offset=offset, dtype=dtype)        
+        
+
+
```

Comment:
Yes this is what I had in mind.
And yes I think the ringbuffer is just a special case a NDRingbuffer.

For me all the code go in the good way in RingBuffer, we just need to extend ND case in RingBuffer by playing on the dimenssion 0 in new_chunk and get_data.

But overall, we need  we need a swapaxes kargs which is None by default. (timeaxis=0)

In the case of (the case i want) swapaxes is (0,1) (timeaxis=1 internally), just after creating the np.array we perform aswapaxes in the __init__ so that when we have a new chunk working on the dimension 0, it is silently done on timeaxis 1 by numpy magic index and strides.



And the goal is this one:
Case continous by sample:
```
length = 1024
nb_channel = 16
ringbuf = NDRingBuffer(shape=(length, nb_channel), double=True, swapaxis=None)
# internally it is
self.buf = np.ones(nb_channel, length)

# so that when

chunk = ringbuf.get_data(start, stop)
chunk[:, chan] # this is not continuous in memory!!!!!!
chunk[pos, :] # this is continuous in memory
```



Case continous by channel:
```
length = 1024
nb_channel = 16
ringbuf = NDRingBuffer(shape=(length, nb_channel), double=True, swapaxis=(0,1))
# internally it is
self.buf = np.ones(nb_channel, length)  # note that there is little hack on the shape which is also swaped
self.buf = self.buf.swapaxis(0,1)

# so that when

chunk = ringbuf.get_data(start, stop)
chunk[:, chan] # this is continuous in memory!!!!!!
chunk[pos, :] # this is not continuous in memory
```

Really sorry to not be clear. I guess this is partly due to my english but you know my level...